src/System.Net.Sockets/tests/FunctionalTests/UnixDomainSocketTest.cs

```diff
@@ -0,0 +1,221 @@
+        }
+
+        [Fact]
+        [PlatformSpecific(PlatformID.Linux | PlatformID.OSX)]
+        public void Socket_ConnectAsyncUnixDomainSocketEndPoint_Success()
+        {
+            Assert.True(Capability.UnixDomainSocketsSupport());
+
+            SocketTestServer server = SocketTestServer.SocketTestServerFactory(_endPoint, ProtocolType.Unspecified);
+
+            SocketAsyncEventArgs args = new SocketAsyncEventArgs();
+            args.RemoteEndPoint = _endPoint;
+            args.Completed += OnConnectAsyncCompleted;
+
+            ManualResetEvent complete = new ManualResetEvent(false);
+            args.UserToken = complete;
+
+            Socket sock = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
+            Assert.True(sock.ConnectAsync(args));
+
+            complete.WaitOne();
+
+            Assert.Equal(SocketError.Success, args.SocketError);
+            Assert.Null(args.ConnectByNameError);
+
+            complete.Dispose();
+            sock.Dispose();
+            server.Dispose();
+        }
+
+        [Fact]
+        [PlatformSpecific(PlatformID.Linux | PlatformID.OSX)]
+        public void Socket_ConnectAsyncUnixDomainSocketEndPoint_NotServer()
+        {
+            Assert.True(Capability.UnixDomainSocketsSupport());
+
+            SocketAsyncEventArgs args = new SocketAsyncEventArgs();
+            args.RemoteEndPoint = _endPoint;
+            args.Completed += OnConnectAsyncCompleted;
+
+            ManualResetEvent complete = new ManualResetEvent(false);
+            args.UserToken = complete;
+
+            Socket sock = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
+
+            bool willRaiseEvent = sock.ConnectAsync(args);
+            if (willRaiseEvent)
+            {
+                complete.WaitOne();
+            }
+
+            Assert.Equal(SocketError.SocketError, args.SocketError);
+
+            complete.Dispose();
+            sock.Dispose();
+        }
+
+        public void Dispose()
+        {
+            if (File.Exists(_path))
+            {
+                try
+                {
+                    File.Delete(_path);
+                }
+                catch
+                {
+                }
+            }
+
+        }
+
+        #region GC Finalizer test
+
+        // This test assumes sequential execution of tests and that it is going to be executed after other tests
+        // that used Sockets. 
+        [Fact]
+        public void TestFinalizers()
+        {
+            // Making several passes through the FReachable list.
+            for (int i = 0; i < 3; i++)
+            {
+                GC.Collect();
+                GC.WaitForPendingFinalizers();
+            }
+        }
+
+        #endregion
+
+        private class UnixDomainSocketEndPoint : EndPoint
+        {
+            private static readonly Encoding PathEncoding = Encoding.UTF8;
```

Comment:
According to http://michal.kosmulski.org/computing/articles/linux-unicode.html, Linux expects a file path to be UTF8. According to http://apple.stackexchange.com/questions/10476/how-to-enter-special-characters-so-that-bash-terminal-understands-them, same applies to OSX.  I did not test it with non-ASCII encodings because we're not shipping UnixDomainSocketEndPoint, it's only used in the tests.